### DISCUSSION TITLE
Why did my Neovim F# tree-sitter config break?

### DISCUSSION THREAD

**Discussion by mattgallagher92** [ionide/tree-sitter-fsharp#82] [Q&A] [answered]
I was facing some issues over the last few days when using Neovim to work with F# files. In particular, I was seeing errors like this:

```
Error in decoration provider treesitter/highlighter.win:
Error executing lua: ...2823/usr/share/nvim/runtime/lua/vim/treesitter/query.lua:252: Query error at 7:3. Invalid node type
 "xml_doc":
((xml_doc (xml_doc_content) @injection.content)
  ^

stack traceback:
        [C]: in function '_ts_parse_query'
        ...2823/usr/share/nvim/runtime/lua/vim/treesitter/query.lua:252: in function 'fn'
        ...im/2823/usr/share/nvim/runtime/lua/vim/func/_memoize.lua:58: in function 'fn'
        ...im/2823/usr/share/nvim/runtime/lua/vim/func/_memoize.lua:58: in function 'get'
        ...r/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:128: in function 'new'
        ...r/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:511: in function 'add_child'
        ...r/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:395: in function '_add_injections'
        ...r/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:447: in function 'parse'
        ...sr/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:397: in function <...sr/share/nvim/runtime/lua/vim/tree
sitter
```

Eventually I realised that perhaps updating the tree-sitter query files might help. I noticed that `xml_doc` was removed in 2f93ed6ac40f79cfb9ba27f5fbc68a988b5bcec1, which was an encouraging sign. Once I replaced the query files with the new versions, things worked again 🥳 

I'm still pretty new to Neovim and tree-sitter, so I'm not certain exactly how a setup that was working broke. My guess is the following: I updated to a newer version of `nvim-treesitter`; that version had breaking changes, so the query files in this repo that worked with an old version of `nvim-treesitter` no longer worked; you updated the query files to work with the new version of `nvim-treesitter`; when I copied them into my config, they were compatible with the version of `nvim-treesitter` that I had installed, so F# tree-sitter integration worked again.

Can you confirm that that's likely what happened? In particular, were there breaking changes in a recent update to `nvim-treesitter` that meant `xml_doc` was no longer a valid node type? A better understanding will help me (and hopefully others) resolve similar situations in future.

**Comment by Nsidorenco** [comment, marked as answer]
The breaking changes are entirely on me 😄 

The node types are defined by the grammar and must match the node types used in the queries. 
If a query references a node that the grammar does not expose, then you get an error like the one your posted.

As to how this happened: When ever you update `nvim-treesitter` it run `:TSUpdate` which updates all parsers to their latest version. If there has been a breaking change to the grammar since you last updated, then your queries will break.

To avoid getting surprised by this you can pin the revision of the parser like so:

```lua
local parser_config = require('nvim-treesitter.parsers').get_parser_configs()
parser_config.fsharp = {
  install_info = {
    url = 'https://github.com/ionide/tree-sitter-fsharp',
    branch = 'main',
    revision = 'some git sha',
    files = { 'src/scanner.c', 'src/parser.c' },
  },
  requires_generate_from_grammar = false,
  filetype = 'fsharp',
}
```

**Reply by mattgallagher92** [reply]
Thanks for the explanation! It's helpful to get a more solid understanding of how everything ties together 🙂

Pinning to a revision sounds like a good idea to ensure that the grammar and queries remain in sync!

Thanks again for all your excellent work making tree-sitter work with F# 😀
